lib/data_provider.js

```diff
@@ -0,0 +1,150 @@
+/* Copyright 2021 F5 Networks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const fs = require('fs');
+
+const ResourceCache = require('./resource_cache').ResourceCache;
+
+/**
+ * Abstract base class for DataProvider classes
+ */
+class BaseDataProvider {
+    constructor() {
+        if (new.target === BaseDataProvider) {
+            throw new TypeError('Cannot instantiate Abstract BaseDataProvider');
+        }
+
+        const abstractMethods = [
+            '_loadData',
+            'list'
+        ];
+        abstractMethods.forEach((method) => {
+            if (this[method] === undefined) {
+                throw new TypeError(`Expected ${method} to be defined`);
+            }
+        });
+
+        this.cache = new ResourceCache(dataName => this._loadData(dataName));
+    }
+
+    /**
+     * Get the data file contents associated with the supplied key
+     *
+     * @param {string} key
+     * @returns {object}
+     */
+    fetch(key) {
+        return this.cache.fetch(key);
+    }
+}
+
+/**
+ * DataProvider that fetches data from the file system
+ */
+class FsDataProvider extends BaseDataProvider {
+    /**
+     * @param {string} dataRootPath - a path to a directory containing data files
+     */
+    constructor(dataRootPath) {
+        super();
+
+        this.data_path = dataRootPath;
```

Comment:
:nit: camelCase for consistency

Reply:
Fixed. I also fixed up a similar issue with SchemaProvider (used an alias to maintain compatibility since the property is technically public).